docs/table-migration.md

```diff
@@ -0,0 +1,89 @@
+ - this work for additional information regarding copyright ownership.
+ - The ASF licenses this file to You under the Apache License, Version 2.0
+ - (the "License"); you may not use this file except in compliance with
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+# Table Migration
+Apache Iceberg supports converting existing tables in other formats to Iceberg tables. This section introduces the general concept of table migration, its approaches, and existing implementations in Iceberg.
+
+## Migration Approaches
+There are two main approaches to perform table migration: CTAS (Create Table As Select) and in-place migration.
+
+### Create-Table-As-Select Migration
+CTAS migration involves creating a new Iceberg table and copying data from the existing table to the new one. This method is preferred when you want to completely cut ties with your old table, ensuring the new table is independent and fully managed by Iceberg.
+However, CTAS migration may require more time to complete and might not be suitable for production use cases where downtime is not acceptable.
+![CTAS Migration](../../../img/iceberg-CTAS.png)
+### In-Place Migration
+In-place migration retains the existing data files but adds Iceberg metadata on top of them. This approach is faster and does not require copying data, making it more suitable for production use cases.
+![In-Place Migration](../../../img/iceberg-In-place.png)
+## In-Place Migration Actions
+Apache Iceberg primarily supports the in-place migration approach, which includes three important actions:
+
+1. Snapshot Table
+2. Migrate Table
+3. Add Files
+
+### Snapshot Table
+The Snapshot Table action creates a new iceberg table with the same schema and partitioning as the source table, leaving the source table unchanged during and after the action.
+
+**Step 1:** Create a new Iceberg table with the same metadata (schema, partition spec, etc.) as the source table
+![Snapshot Table Step 1](../../../img/iceberg-snapshotaction-step1.png)
+
+**Step 2:** Commit all data files across all partitions to the new Iceberg table. The source table remains unchanged.
+![Snapshot Table Step 2](../../../img/iceberg-snapshotaction-step2.png)
+### Migrate Table
+The Migrate Table action also creates a new Iceberg table with the same schema and partitioning as the source table. However, during the action execution, it locks and drops the source table from the catalog.
+Consequently, Migrate Table requires all readers and writers working on the source table to be stopped before the action is performed.
```

Comment:
I think for writers you are right they need to all stop, for reader it does not have to be, because the reader can support reading both types of table.

Reply:
Thank you for your explanation. I have a follow-up question regarding the migration process. 

When performing the migration, I understand that the source table is first renamed as a backup, and a staged iceberg table with the same identifier is created. However, it appears that the staged changes will not be committed until all data files have been migrated. This creates a period during the migration process where readers will not be able to find any table with the source table identifier in the catalog. In light of this, I would like to confirm if it is still safe to keep the existing readers active during this period. Thank you!